runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -359,16 +534,86 @@ public void finishBundle() {
         emitResults();
       } catch (Exception e) {
         throw new RuntimeException("Failed to finish remote bundle", e);
+      } finally {
+        remoteBundle = null;
+      }
+      if (bundleFinishedCallback != null) {
+        bundleFinishedCallback.run();
+        bundleFinishedCallback = null;
       }
     }
 
+    void setTimerKey(Object key) {
+      this.beforeFireTimerKey = key;
+    }
+
+    boolean isBundleInProgress() {
+      return remoteBundle != null;
+    }
+
+    void setBundleFinishedCallback(Runnable callback) {
+      this.bundleFinishedCallback = callback;
+    }
+
     private void emitResults() {
       KV<String, OutputT> result;
       while ((result = outputQueue.poll()) != null) {
-        outputManager.output(outputMap.get(result.getKey()), (WindowedValue) result.getValue());
+        final String inputCollectionId = result.getKey();
+        TupleTag<?> tag = outputMap.get(inputCollectionId);
+        WindowedValue windowedValue =
+            Preconditions.checkNotNull(
+                (WindowedValue) result.getValue(),
+                "Received a null value from the SDK harness for %s",
+                inputCollectionId);
+        if (tag != null) {
+          // process regular elements
+          outputManager.output(tag, windowedValue);
+        } else {
+          // process timer elements
+          // TODO This is ugly. There should be an easier way to retrieve the
+          String timerPCollectionId =
+              inputCollectionId.substring(0, inputCollectionId.length() - ".out:0".length());
+          TimerReference timerReference = timerReferenceMap.get(timerPCollectionId);
+          if (timerReference != null) {
+            Timer timer =
+                Preconditions.checkNotNull(
+                    (Timer) ((KV) windowedValue.getValue()).getValue(),
+                    "Received null Timer from SDK harness: %s",
+                    windowedValue);
+            LOG.debug("Timer received: {} {}", inputCollectionId, timer);
+            for (Object window : windowedValue.getWindows()) {
+              StateNamespace namespace =
+                  StateNamespaces.window(windowCoder, (BoundedWindow) window);
+              TimerSpec timerSpec = extractTimerSpec(timerReference);
+              TimerInternals.TimerData timerData =
+                  TimerInternals.TimerData.of(
+                      timerPCollectionId,
+                      namespace,
+                      timer.getTimestamp(),
+                      timerSpec.getTimeDomain());
+              timerDataConsumer.accept(windowedValue, timerData);
```

Comment:
This is another example of unnecessary code complexity. Had to navigate a chain in the IDE when this otherwise could just be a simple method call on `ExecutableStageDoFnOperator`.

Reply:
I agree it can be confusing. On the other hand, this decouples the timer generation code and the timer consumer code.